### DISCUSSION TITLE
containerd support container groups per shim,how to use Docker support in Kubernetes

### DISCUSSION THREAD

**Discussion by gosoon** [moby/moby#43962] [Q&A] [answered]
### Description

Containerd support [io.containerd.runc.v2 with Container Groups](https://github.com/containerd/containerd/pull/3004#top), If Docker can support it, it will reduce memory usage. I wonder if Docker supports configuration.

**Comment by cpuguy83** [comment, marked as answer]
Kubernetes has the concept of pods where by containers have a shared(-ish) lifecycle.
There is no such concept in Docker and without that using groups doesn't really make sense.

In the upcoming release we support using custom shims with docker.
If your desire is to reduce memory consumption there are a couple of things you can use in that case:

1. There is a [rust version](https://github.com/containerd/rust-extensions/tree/main/crates/runc-shim) of the runc shim which , although I haven't messured the shim itself, should have lower memory usage.
2. There is is a [systemd backed shim](https://github.com/cpuguy83/containerd-shim-systemd-v1) (Still uses runc under the hood) which is a single shim process for the entire node
3. [embed shim](https://github.com/fuweid/embedshim) is a shim which does not require a shim daemon and instead uses ebpf to monitor containers.

**Reply by gosoon** [reply]
I mean I want to use container groups per shim under the default shim,kubelet use docker or containerd as container runtime,also use default containerd shim(io.containerd.runc.v2),default containerd shim(io.containerd.runc.v2) already support group containers under a single shim.

As the [io.containerd.runc.v2 with Container Groups](https://github.com/containerd/containerd/pull/3004#top) says,if k8s pod using the following labels:

- `io.containerd.runc.v2.group=<group>`
- `io.kubernetes.cri.sandbox-id=<group>`

and create container using `Annotations` on the runtime spec can support group multiple containers and only launch one shim for this group of containers.

After my tests,if kubelet use containerd as its container runtime,and creating the pod, the created containers under a single shim,if kubelet use dockershim create pod with `io.kubernetes.cri.sandbox-id` label,  the created containers not support group containers under a single shim.

After looking through the Moby code,I find create container not add  `io.kubernetes.cri.sandbox-id` label to `Annotations` on the runtime spec.

**Reply by cpuguy83** [reply]
Right, docker does not have a pod concept.
Docker is not designed to be used the way Kubernetes used/uses it.
Why not have Kubernetes use containerd?

**Reply by gosoon** [reply]
Thank you for your reply. I understand. I always thought Docker was supportive, but I was mistaken.
